tools/cldr-code/src/main/java/org/unicode/cldr/util/personname/PersonNameFormatter.java

```diff
@@ -0,0 +1,580 @@
+
+        public NamePattern getBestMatch(NameObject nameObject, FormatParameters nameFormatParameters) {
+            if (nameFormatParameters.order == null) {
+                nameFormatParameters = nameFormatParameters.setOrder(localeToOrder.get(nameObject.getNameLocale()));
+            }
+
+            // Sift through the options to get the first one that best matches nameObjectFields
+
+            for (Entry<ParameterMatcher, Collection<NamePattern>> parametersAndPatterns : parameterMatcherToNamePattern.asMap().entrySet()) {
+                ParameterMatcher parameters = parametersAndPatterns.getKey();
+                if (!parameters.matches(nameFormatParameters)) {
+                    continue;
+                }
+
+                Collection<NamePattern> namePatterns = parametersAndPatterns.getValue();
+
+                // TODO pick the NamePattern that best matches the fields in the nameObject
+                // for now, just return the first
+
+                return namePatterns.iterator().next();
+            }
+            return null; // for now; this will only happen if the NamePatternData is invalid
+        }
+
+        /**
+         * TODO Replace by builder. Eventually will be built from CLDR XML data.
+         * The multimap values must retain the order they are built with!
+         */
+        public NamePatternData(ImmutableMap<ULocale, Order> localeToOrder, ImmutableListMultimap<ParameterMatcher, NamePattern> formatParametersToNamePattern) {
+            this.localeToOrder = localeToOrder == null ? ImmutableMap.of() : localeToOrder;
+            this.parameterMatcherToNamePattern = formatParametersToNamePattern;
+            // TODO: check formatParametersToNamePattern for validity
+            // * no null values
+            // * no ParameterMatcher should be completely masked by previous ones
+            // * each entry in ParameterMatcher must have at least one NamePattern
+            // * the final entry must have MATCH_ALL as the key
+        }
+
+        // TODO: add method that takes a string instead of a NamePatten
+
+        @Override
+        public String toString() {
+            return "{" + (localeToOrder.isEmpty() ? "" : "localeToOrder=" + localeToOrder + "\n\t\t")
+                + show(parameterMatcherToNamePattern) + "}";
+        }
+
+        private String show(ImmutableListMultimap<ParameterMatcher, NamePattern> multimap) {
+//            StringBuilder result = new StringBuilder();
+//            for (Entry<ParameterMatcher, Collection<NamePattern>> entry : multimap.asMap().entrySet()) {
+//                ParameterMatcher key = entry.getKey();
+//                if
+//            }
+            return parameterMatcherToNamePattern.asMap().toString();
+        }
+    }
+
+    public static interface NameObject {
+        public String getBestValue(ModifiedField modifiedField, Set<Modifier> unhandledModifiers);
```

Comment:
I do not understand why the arg is called `unhandledModifiers`.

The spec says:

_"The name formatter queries the name object for each possible combination of the specified modifiers, and if any one of the combinations produces a result, then that result is used (with any remaining modifiers applied algorithmically)."_

The way I read it, the arg should be called `modifiers`. As formatter implementation churns through combinations, it creates its own set of "unhandled" modifiers, which it then uses to apply default logic. In other words, from the `NameObject`'s perspective this arg is not "unhandled", but "requested" modifiers; it just may not handle what's requested.

Reply:
I think the spec provides defaults for handling outside of the NameObject.